### ISSUE TITLE
Distinguish loading and transparent WMS Titles

### ISSUE THREAD

**Issue by SlowMo24** [openlayers/openlayers#14500]
**Problem**

I'm counting the number of pixels per color in an `ol.source.Raster` using an `ol.source.TileWMS` as source. But I cannot distinguish between loading tiles and NoData tiles which are both fed to my `operation`-function as (0,0,0,0). Yet, the WMS server sends out (255,255,255,0) as NoData.

I have solved the issue by adding `params: {"TRANSPARENT": false}` for the `TileWMS` source of my raster. Unfortunately this means that I have to make a second call to the WMS server to get a transparent image for map display.

**Desired solution**

Do not zero-fill transparent images or add an option to prevent it. I think this is related to #13570 .

**Comment by ahocevar** [comment]
If it is enough to count pixels when you know the TileWMS's tiles have finished loading, you could listen for the map's `loadstart` and `loadend` events, and refrain from counting pixels between the two.

**Comment by mike-000** [comment]
> Do not zero-fill transparent images or add an option to prevent it. I think this is related to #13570 .

Alpha pre-multiplication is unavoidable when using 2d canvases.  Additionally with IOS it cannot be prevented in WebGL. To obtain the original values from a PNG you would need a third party PNG parser as in https://github.com/openlayers/openlayers/issues/13934#issuecomment-1214217865

**Comment by SlowMo24** [comment]
thank you both for your comments and even coding a solution!

Two questions:

Would it also be possible to initialise the image fed to the Raster by the TileWMS with different values (e.g. white) or does that have no effect/side effects? I could not find the location where the image is initialised.

@ahocevar if we listened to these events, how would we get access to the pixel data or trigger a final `operation` on the Raster after `loadend`?

**Comment by ahocevar** [comment]
@SlowMo24 When `loadend` is triggered, you can use the next (final) render cycle to do your calculation, probably with your existing code. Just skip the calculation between `loadstart` and `loadend` of the map.

**Comment by mike-000** [comment]
> Would it also be possible to initialise the image fed to the Raster by the TileWMS with different values (e.g. white) or does that have no effect/side effects? I could not find the location where the image is initialised.

Yes, when using a DataTile source you can and should test for transparent white and style the rendered output to opaque white or any other unique color which suits your application.  If left as transparent white it would be pre-multiplied to transparent black by the Raster source canvas making it indistinguishable again.  If changed to opaque white the Raster source operation could count the white pixels before resetting them to transparent for the final output if that is required.  https://codesandbox.io/s/simple-forked-7xgwn2?file=/main.js
